### PR TITLE
feat(core) make service on routes optional (transparent proxying)

### PR DESCRIPTION
**Pull request by bungle**
### Summary

This is first implementation of so called `allow_by_default` routing. E.g. routes that act as a transparent routes. Which means you don't need to create service entity, but instead you can use HTTP proxying or `iptables` transparent proxying rules.

### CODE REVIEW

**Comment by bungle** [comment]
@james-callahan can you do a small feasability check on this? Am I on a right page, or do you see some problems? If everything is fine, I think I can move to adding tests.

File: kong/runloop/handler.lua [hunk 1]
```diff
@@ -77,34 +78,49 @@ do
         return nil, "could not load routes: " .. err
       end
 
-      local service_pk = route.service
-
-      if not service_pk then
-        return nil, "route (" .. route.id .. ") is not associated with service"
+      local process_route
+      for _, protocol in ipairs(route.protocols) do
+        if protocol_subsystem[protocol] == subsystem then
```
[the remark this answers is not in the document]
Comment (bungle) [reply kong/runloop/handler.lua]: Replaced.

File: kong/router.lua [hunk 2]
```diff
@@ -1229,11 +1236,14 @@ function _M.new(routes)
...
-        ngx.header["Kong-Route-Id"]   = match_t.route.id
-        ngx.header["Kong-Service-Id"] = match_t.service.id
+        ngx.header["Kong-Route-Id"] = match_t.route.id
+
+        if match_t.service then
+          ngx.header["Kong-Service-Id"] = match_t.service.id
 
-        if match_t.service.name then
-          ngx.header["Kong-Service-Name"] = match_t.service.name
+          if match_t.service.name then
```
[the remark this answers is not in the document]
Comment (bungle) [reply kong/router.lua]: Protected that and the route as well.

File: kong/runloop/handler.lua [hunk 3]
```diff
@@ -737,25 +725,82 @@ return {
...
+      if not service then
+        local service_scheme
+        local service_host
+        local service_port
+
+        local http_host = var.http_host
+        if http_host then
+          http_host = lower(http_host)
+
+          local s = find(http_host, ":", 2, true)
```
[the remark this answers is not in the document]
Comment (bungle) [reply kong/runloop/handler.lua]: Added comment above.

File: kong/router.lua [hunk 4]
```diff
@@ -323,38 +330,41 @@ local function marshall_route(r)
...
-  end
+  -- upstream_url parsing
 
 
-  -- upstream_url parsing
+  if protocol then
+    route_t.upstream_url_t.scheme = protocol
+  end
 
+  do
```
Comment (james-callahan) [inline kong/router.lua]: Why the `do/end` block here?

File: kong/runloop/handler.lua [hunk 5]
```diff
@@ -721,25 +718,86 @@ return {
 
       local trusted_ip = kong.ip.is_trusted(realip_remote_addr)
       if trusted_ip then
-        forwarded_proto = var.http_x_forwarded_proto or var.scheme
-        forwarded_host  = var.http_x_forwarded_host  or var.host
-        forwarded_port  = var.http_x_forwarded_port  or var.server_port
+        forwarded_proto = var.http_x_forwarded_proto          or scheme
+        forwarded_host  = var.http_x_forwarded_host           or host
+        forwarded_port  = tonumber(var.http_x_forwarded_port) or port
```
Comment (james-callahan) [inline kong/runloop/handler.lua]: Pass a base of `10` to `tonumber`?

File: kong/runloop/handler.lua [hunk 6]
```diff
@@ -721,25 +718,86 @@ return {
...
+        if http_host then
+          http_host = lower(http_host)
+
+          -- Micro-optimization to make sure we use JITable functions in this
+          -- potentially hot path, e.g. this simple code doesn't warrant us
+          -- using something like socket.url.parse:
+          -- https://github.com/diegonehab/luasocket/blob/043e99771352aff47680b99f09b66a32f0cc3ef5/src/url.lua#L143-L192
+          local s = find(http_host, ":", 2, true)
+          if s then
+            local p = tonumber(sub(http_host, s + 1))
```
Comment (james-callahan) [inline kong/runloop/handler.lua]: Pass base to `tonumber`?

File: kong/router.lua (hunk 4 above)
Comment (bungle) [reply kong/router.lua]: Removed.

File: kong/runloop/handler.lua (hunk 5 above)
Comment (bungle) [reply kong/runloop/handler.lua]: Passing it now.

File: kong/runloop/handler.lua (hunk 6 above)
Comment (bungle) [reply kong/runloop/handler.lua]: Passing it now. Thank you @james-callahan for the review!

File: kong/router.lua [hunk 7]
```diff
@@ -1150,7 +1157,7 @@ function _M.new(routes)
               else
                 -- we retain the incoming path, just prefix it with the upstream
                 -- path, but skip the initial slash
-                upstream_uri = upstream_base .. sub(req_uri, 2, -1)
+                upstream_uri = upstream_base .. sub(req_uri, 2)
```
Comment (james-callahan) [inline kong/router.lua]: Superficial change intended?

Comment (bungle) [reply kong/router.lua]: Reverted that.

File: kong/runloop/handler.lua [hunk 8]
```diff
@@ -721,25 +718,86 @@ return {
...
+        if http_host then
+          http_host = lower(http_host)
+
+          -- Micro-optimization to make sure we use JITable functions in this
+          -- potentially hot path, e.g. this simple code doesn't warrant us
+          -- using something like socket.url.parse:
+          -- https://github.com/diegonehab/luasocket/blob/043e99771352aff47680b99f09b66a32f0cc3ef5/src/url.lua#L143-L192
+          local s = find(http_host, ":", 2, true)
+          if s then
+            local p = tonumber(sub(http_host, s + 1), 10)
```
Comment (james-callahan) [inline kong/runloop/handler.lua]: Make sure you check for a non-whole number e.g. user provides `1000.5`

File: kong/runloop/handler.lua [hunk 9]
```diff
@@ -721,25 +718,86 @@ return {
...
+            service_port = var.server_port
+          end
+        end
+
+        if service_port == 443 then
+          service_scheme = "https"
+        elseif service_port == 80 then
+          service_scheme = "http"
+        else
+          service_scheme = scheme
```
Comment (james-callahan) [inline kong/runloop/handler.lua]: Shouldn't we always set `service_scheme` to `scheme`?

File: kong/runloop/handler.lua (hunk 8 above)
Comment (bungle) [reply kong/runloop/handler.lua]: Added check (one line below this)

Comment (bungle) [reply kong/runloop/handler.lua]: Now there is a problem with `10.` too :-)

Comment (bungle) [reply kong/runloop/handler.lua]: And also scientific notation like `12e3`.

File: kong/runloop/handler.lua (hunk 9 above)
Comment (bungle) [reply kong/runloop/handler.lua]: I am not sure, let's say you connect to Kong using `https` and have a `host` header of `Host: www.google.com:80`, Should we still connect to `www.google.com` on port `80` with `http` or `https`?

Comment (james-callahan) [reply kong/runloop/handler.lua]: I don't think we should add special cases here: simpler to just have a rule "scheme in is scheme out". 

Note that in future we plan on having a route property like "terminate_tls" (recall https://github.com/Kong/kong/blob/aa034cda1be49a05751cb2695b1b56bbbcfafd1e/kong/runloop/handler.lua#L627), using that would probably be the right place to let a user configure http vs https.

File: kong/runloop/handler.lua [hunk 10]
```diff
@@ -721,25 +724,114 @@ return {
...
+        local service_host
+        local service_port
+
+        -- 1. try to find information from a request-line
+        if var.http2 == "" then
+          local request_line = var.request
+          if request_line then
+            local matches = ngx.re.match(request_line, [[\w+ (\S+)]], "ajos")
+            if matches and matches[1] then
+              local uri = url.parse(matches[1])
```
Comment (james-callahan) [inline kong/runloop/handler.lua]: Could this be done with a regex that extracts matches?

I think this could be a hot path. But as it's new functionality we could just fix it later.
maybe add a `-- TODO: optimize`?

File: kong/runloop/handler.lua [hunk 11]
```diff
@@ -721,25 +724,114 @@ return {
...
+              local uri_scheme = uri and uri.scheme
+              if uri_scheme then
+                uri_scheme = lower(uri_scheme)
```
Comment (james-callahan) [inline kong/runloop/handler.lua]: Interesting: I this should be done by luasocket; I think the lack of it should be considered a luasocket bug.

File: kong/runloop/handler.lua [hunk 12]
```diff
@@ -721,25 +724,114 @@ return {
...
+                if uri_scheme ~= "http" or uri_scheme ~= "https" then
```
Comment (james-callahan) [inline kong/runloop/handler.lua]: For a websocket connection, do we need to check for `"ws"` and `"wss"`?

File: kong/runloop/handler.lua [hunk 13]
```diff
@@ -21,15 +22,18 @@ local certificate = require "kong.runloop.certificate"
...
 local tonumber    = tonumber
 local sub         = string.sub
+local find        = string.find
 local lower       = string.lower
 local fmt         = string.format
 local sort        = table.sort
 local ngx         = ngx
 local log         = ngx.log
 local ngx_now     = ngx.now
+local rematch     = ngx.re.match
```
Comment (james-callahan) [inline kong/runloop/handler.lua]: Is this a conventional alias for this function? I would have expected `re_match`. (I got confused when reading further down)

File: kong/runloop/handler.lua (hunk 11 above)
Comment (bungle) [reply kong/runloop/handler.lua]:
```
> = require "socket.url".parse("HttP://www.google.com").scheme
HttP
```

Comment (james-callahan) [reply kong/runloop/handler.lua]: Yes. I would consider that a luasocket bug.

Fine to work around it as you have.

File: kong/runloop/handler.lua [hunk 14]
```diff
@@ -77,34 +74,49 @@ do
...
+        local service_subsystem
+
+        if service_pk then
+          service, err = db.services:select(service_pk)
+          if not service then
+            return nil, "could not find service for route (" .. route.id .. "): " ..
+                        err
+          end
+
+          service_subsystem = constants.PROTOCOLS_WITH_SUBSYSTEM[service.protocol]
```
[the remark this answers is not in the document]
Comment (bungle) [reply kong/runloop/handler.lua]: This is possible ATM:
```
 http PUT :8001/services/test url=http://www.google.com/
http POST :8001/services/test/routes destinations[1].ip=0.0.0.0/0 protocols=tcp -f 
```

[the remark this answers is not in the document]
Comment (james-callahan) [reply kong/runloop/handler.lua]: Ah. It's not meant to be. We ran out of time to add the restriction for this sort of thing to the metaschema. @hishamhm probably recalls the conversation?

So I guess leave this branch in here for now, but perhaps leave a TODO saying "should be enforced by schema"?

File: kong/db/schema/entities/routes.lua [hunk 15]
```diff
@@ -104,7 +104,7 @@ return {
                        },
                      }, },
     { tags             = typedefs.tags },
-    { service = { type = "foreign", reference = "services", required = true }, },
+    { service = { type = "foreign", reference = "services", required = false }, },
```
[the remark this answers is not in the document]
Comment (hishamhm) [reply kong/db/schema/entities/routes.lua]: I'd prefer leaving it out, but no big issue

[the remark this answers is not in the document]
Comment (hishamhm) [reply kong/db/schema/entities/routes.lua]: (maybe it could be confusing for newcomers though)

File: kong/runloop/handler.lua (hunk 14 above)
[the remark this answers is not in the document]
Comment (hishamhm) [reply kong/runloop/handler.lua]: You mean doing something similar to what @kikito did in 608872dbfbcc1dbaec6e680f7dac52160ff0de73 for checking that protocols in plugins and routes match?

Do we need something similar for services and routes?

File: kong/db/schema/entities/routes.lua (hunk 15 above)
[the remark this answers is not in the document]
Comment (bungle) [reply kong/db/schema/entities/routes.lua]: removed.

File: kong/runloop/handler.lua (hunk 10 above)
Comment (bungle) [reply kong/runloop/handler.lua]: I rewrote that without `socket.url`.

File: kong/runloop/handler.lua (hunk 11 above)
Comment (bungle) [reply kong/runloop/handler.lua]: Worked this to remove dependency on `socket.url`.

File: kong/runloop/handler.lua (hunk 12 above)
Comment (bungle) [reply kong/runloop/handler.lua]: It now checks for `ws` and `wss` too.

File: kong/runloop/handler.lua (hunk 13 above)
Comment (bungle) [reply kong/runloop/handler.lua]: changed to `re_match`.

File: kong/runloop/handler.lua (hunk 14 above)
[the remark this answers is not in the document]
Comment (bungle) [reply kong/runloop/handler.lua]: @hishamhm so yes, I think so. I am not sure can we downgrade `http/https` to `tcp/tls`, and I am not sure if we can upgrade `tcp/tls` to `http/https`, at least we don't route such (I didn't invent `service.protocol` check here in router rebuild, I just added route `subsystem` check, which is quite obvious thing to do as there we don't have to deal with such question anyway.

File: kong/runloop/handler.lua [hunk 16]
```diff
@@ -77,34 +74,49 @@ do
         return nil, "could not load routes: " .. err
       end
 
-      local service_pk = route.service
-
-      if not service_pk then
-        return nil, "route (" .. route.id .. ") is not associated with service"
+      local process_route
```
[the remark this answers is not in the document]
Comment (bungle) [reply kong/runloop/handler.lua]: The helper was added.

File: kong/runloop/handler.lua (hunk 14 above)
[the remark this answers is not in the document]
Comment (james-callahan) [reply kong/runloop/handler.lua]: @hishamhm yes.

File: kong/runloop/handler.lua [hunk 17]
```diff
@@ -59,16 +62,37 @@ end
...
-
   local router
   local router_version
 
+  local function should_process_route(route)
+    for _, protocol in ipairs(route.protocols) do
+      if constants.PROTOCOLS_WITH_SUBSYSTEM[protocol] == subsystem then
+        return true
+      end
+    end
```
Comment (james-callahan) [inline kong/runloop/handler.lua]: Add `return false` rather than falling off?

File: kong/runloop/handler.lua [hunk 18]
```diff
@@ -59,16 +62,37 @@ end
...
+
+    local service, err = db.services:select(service_pk)
+    if not service then
+      return nil, "could not find service for route (" .. route.id .. "): " ..
+                  err
+    end
+
+    -- TODO: this should not be needed as the schema should check it already
+    local service_subsystem = constants.PROTOCOLS_WITH_SUBSYSTEM[service.protocol]
+    if service_subsystem == subsystem then
```
Comment (james-callahan) [inline kong/runloop/handler.lua]: I'd prefer to invert this check and return a good error message.
That way we `return` for the happy case

File: kong/runloop/handler.lua [hunk 19]
```diff
@@ -721,25 +728,117 @@ return {
...
+        if var.http2 == "" then
+          local request_line = var.request
+          if request_line then
+            local matches = re_match(request_line, [[\w+ (https?|wss?)://([^/?#\s]+)]], "ajos")
+            if matches then
+              local uri_scheme = lower(matches[1])
+              if uri_scheme == "https" or uri_scheme == "http" then
+                service_scheme = uri_scheme
+                service_host   = lower(matches[2])
+              elseif uri_scheme == "wss" then
```
Comment (james-callahan) [inline kong/runloop/handler.lua]: I wasn't 100% sure these paths were required, it was just an idea: did you verify what occurs if a websocket client uses a http proxy?

Comment (bungle) [reply kong/runloop/handler.lua]: @james-callahan, it needs testing. The devil is in details in this PR, so thank you for all the feedback so far. We need to get this right. Currently adding more tests to it.

File: kong/runloop/handler.lua (hunk 18 above)
Comment (bungle) [reply kong/runloop/handler.lua]: @james-callahan 

Do you consider logging it is enough? I mean that if we forcibly `return` the router rebuilding will never work. I know that the db should not be in that state ever, but still, I think it is better to just log it and move on.

File: kong/runloop/handler.lua (hunk 17 above)
Comment (bungle) [reply kong/runloop/handler.lua]: Added that.

File: kong/runloop/handler.lua (hunk 18 above)
Comment (bungle) [reply kong/runloop/handler.lua]: Added logging.

File: kong/runloop/handler.lua (hunk 19 above)
Comment (bungle) [reply kong/runloop/handler.lua]: There is now more tests added. Lets figure out the websocket thing later as I think it needs in most cases `Connect`, and that is what we (or Nginx) don't support currently.

File: kong/runloop/handler.lua (hunk 14 above)
[the remark this answers is not in the document]
Comment (bungle) [reply kong/runloop/handler.lua]: I mark this as resolved here as it is not part of this PR.

File: kong/runloop/handler.lua (hunk 18 above)
Comment (james-callahan) [reply kong/runloop/handler.lua]: logging is good. I'd still like to inverse this `if` statement.

File: kong/runloop/handler.lua [hunk 20]
```diff
@@ -698,25 +710,118 @@ return {
...
+        -- Serviceless HTTP / HTTPS / HTTP2 route
+        -----------------------------------------------------------------------
+        local service_scheme
+        local service_host
+        local service_port
+
+        -- 1. try to find information from a request-line
+        local request_line = var.request
+        if request_line then
+          local matches = re_match(request_line, [[\w+ (https?)://([^/?#\s]+)]], "ajos")
```
Comment (james-callahan) [inline kong/runloop/handler.lua]: does this pattern need anchoring?

File: spec/02-integration/05-proxy/18-serviceless_proxying_spec.lua [hunk 21]
```diff
@@ -0,0 +1,468 @@
...
+
+          assert(helpers.start_kong({
+            database      = strategy,
+            nginx_conf    = "spec/fixtures/custom_nginx.template",
+            stream_listen = "off",
+            admin_listen  = "off",
+          }))
+
+        else
+          yaml_file = helpers.make_yaml_file [[
```
Comment (james-callahan) [inline spec/02-integration/05-proxy/18-serviceless_proxying_spec.lua]: use `unindent` helper?

File: spec/02-integration/05-proxy/18-serviceless_proxying_spec.lua [hunk 22]
```diff
@@ -0,0 +1,468 @@
...
+  - paths:
+    - /
+]]
+
+          assert(helpers.start_kong({
+            database           = strategy,
+            nginx_conf         = "spec/fixtures/custom_nginx.template",
+            declarative_config = yaml_file,
+            stream_listen      = "off",
+            admin_liten        = "off",
```
Comment (james-callahan) [inline spec/02-integration/05-proxy/18-serviceless_proxying_spec.lua]: typo: liten vs listen

File: spec/02-integration/05-proxy/18-serviceless_proxying_spec.lua [hunk 23]
```diff
@@ -0,0 +1,468 @@
...
+      lazy_teardown(function()
+        if yaml_file then
+          os.remove(yaml_file)
+        end
+
+        helpers.stop_kong(helpers.test_conf.prefix, true)
+      end)
+
+      it("proxies http to http (request-line)", function()
+        local request = assert(http_request.new_from_uri(HTTP_UPSTREAM_URI))
```
Comment (james-callahan) [inline spec/02-integration/05-proxy/18-serviceless_proxying_spec.lua]: this doesn't need an assert

File: spec/02-integration/05-proxy/18-serviceless_proxying_spec.lua [hunk 24]
```diff
@@ -0,0 +1,468 @@
...
+        request.proxy = HTTP_PROXY_URI
+        request.version = 1.1
+        request.tls = false
+        local headers, stream = request:go()
+
+        assert.equal(200, tonumber(headers:get(":status")))
+        assert.equal(meta._SERVER_TOKENS, headers:get("via"))
```
Comment (james-callahan) [inline spec/02-integration/05-proxy/18-serviceless_proxying_spec.lua]: I think you need to wrap the `:get` in parenthesis.

File: spec/02-integration/05-proxy/18-serviceless_proxying_spec.lua [hunk 25]
```diff
@@ -0,0 +1,468 @@
...
+        ctx:setVerify(openssl_ssl.VERIFY_NONE)
+
+        local request = assert(http_request.new_from_uri(HTTPS_UPSTREAM_URI))
+        request.proxy = HTTP_PROXY_URI
+        request.version = 1.1
+        request.tls = true
+        request.ctx = ctx
+        local headers, stream = request:go()
+
+        print(stream)
```
Comment (james-callahan) [inline spec/02-integration/05-proxy/18-serviceless_proxying_spec.lua]: leftover print?

File: kong/runloop/handler.lua (hunk 20 above)
Comment (bungle) [reply kong/runloop/handler.lua]: `a` means it is anchored.

File: kong/runloop/handler.lua [hunk 26]
```diff
@@ -58,16 +62,41 @@ end
...
+    if not service then
+      return nil, "could not find service for route (" .. route.id .. "): " ..
+                  err
+    end
+
+    -- TODO: this should not be needed as the schema should check it already
+    local service_subsystem = constants.PROTOCOLS_WITH_SUBSYSTEM[service.protocol]
+    if service_subsystem ~= subsystem then
+      log(WARN, "service with protocol '", service.protocol,
+                "' cannot be used with '", subsystem, "' subsystem")
```
Comment (thibaultcha) [inline kong/runloop/handler.lua]: Should we bail out here? `return` with no error, simply log the warning and ignore the Route. Otherwise we may end up in situations where a particular subsystem's Route needs to Route into another subsystem (http to stream or vice-versa), which we are not ready to handle yet.

Or what's the rationale for keeping the Service attached to the Route, and do we have any tests covering this edge case (I assume not, since the schema should not even allow this, in which case, we are introducing untested/undefined behavior in our core logic).

Additionally, this way, we don't have to return 3 values from this function anymore, and callers of this function may use `subsystem` (the top-chunk upvalue) instead of `service_subsystem`.

File: kong/runloop/handler.lua [hunk 27]
```diff
@@ -698,25 +710,118 @@ return {
...
+          if pos then
+            service_host = sub(service_host, pos + 1)
+          end
+
+          pos = find(service_host, ":", 2, true)
+          if pos then
+            service_port = sub(service_host, pos + 1)
+            service_host = sub(service_host, 1, pos - 1)
+
+            if re_match(service_port, [[[1-9]{1}\d{0,4}$]], "adjo") then
```
[the remark this answers is not in the document]
Comment (thibaultcha) [reply kong/runloop/handler.lua]: perf: using `ngx.re.find` should be cheaper and achieve the same result in this context.

File: kong/runloop/handler.lua (hunk 26 above)
Comment (bungle) [reply kong/runloop/handler.lua]: Yes, it was like that, but @james-callahan wanted it to be inverted and it introduced this bug. Damnit. I'll fix it.

File: kong/runloop/handler.lua [hunk 28]
```diff
@@ -698,25 +710,118 @@ return {
...
-        forwarded_port  = var.server_port
+        forwarded_proto = scheme
+        forwarded_host  = host
+        forwarded_port  = port
       end
 
-      local protocols = route.protocols
-      if (protocols and
-          protocols.https and not protocols.http and forwarded_proto ~= "https")
+      local protocols = route and route.protocols
```
[the remark this answers is not in the document]
Comment (bungle) [reply kong/runloop/handler.lua]: removed the `route and `.

File: kong/runloop/handler.lua (hunk 20 above)
[the remark this answers is not in the document]
Comment (bungle) [reply kong/runloop/handler.lua]: Added error check that logs with warning.

File: kong/runloop/handler.lua (hunk 27 above)
[the remark this answers is not in the document]
Comment (bungle) [reply kong/runloop/handler.lua]: Added error check that logs with warning, and replaced match with find.

File: kong/runloop/handler.lua (hunk 26 above)
Comment (bungle) [reply kong/runloop/handler.lua]: Made it return service OR nil, err OR nil (serviceless route).

File: kong/runloop/handler.lua [hunk 29]
```diff
@@ -698,25 +713,125 @@ return {
...
+          if err then
+            log(WARN, "pcre runtime error when matching a request-line: ", err)
+
+          elseif matches then
+            local uri_scheme = lower(matches[1])
+            if uri_scheme == "https" or uri_scheme == "http" then
+              service_scheme = uri_scheme
+              service_host   = lower(matches[2])
+            end
+            --[[ TODO: check if these make sense here?
```
Comment (hutchic) [inline kong/runloop/handler.lua]: left over comment?

Comment (bungle) [reply kong/runloop/handler.lua]: I wanted to leave them here, maybe we return back to that later. I believe we need `Connect` for that to work, but I am not 100% sure.
